### PR TITLE
NodeGraph: Use layered layout instead of force based layout

### PR DESCRIPTION
**Pull request by aocenas**
**What is this feature?**

Changes the default layout for node graph from a d3 force based simulation to a more predictable and orderly layered layout. It uses https://github.com/microsoft/msagljs library for the layout algorithm. This happens only up to 500 nodes though as the layered layout has worse performance at higher number of nodes.

**Why do we need this feature?**

Comparison of the layouts:
| D3 Force | Layered |
|--------|--------|
| Screenshot 2023-10-23 at 15 49 06 | Screenshot 2023-12-01 at 15 40 18 |
| Screenshot 2023-10-23 at 15 49 27 | Screenshot 2023-12-01 at 15 39 42 | 

As can be seen, a nice property of the layered layout is that it neatly organizes the graph into layers which should help for service graph usecases where such "layer-isation" is more expected. It also better separates distinct clusters as can be seen in the bigger service graph where with the D3 force layout it seems like there are 4 distinct clusters while with layered layout it's apparent there are actually 8.

**Who is this feature for?**

[Add information on what kind of user the feature is for.]

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Comment by domasx2** [comment]
niiice <3

**Review by joey-grafana — COMMENTED** [review]
May not be for this PR but seems there is overlap sometimes in the test data source. Also I'm not sure what effect seed is actually having.

Screenshot 2023-12-11 at 11 15 55

File: public/app/plugins/panel/nodeGraph/layoutMsagl.worker.js [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+  const [newNodes, newEdges] = layout(nodes, edges, config);
+  postMessage({ nodes: newNodes, edges: newEdges });
+});
+
+/**
+ * Use d3 force layout to lay the nodes in a sensible way. This function modifies the nodes adding the x,y positions
+ * and also fills in node references in edges instead of node ids.
+ */
+export function layout(nodes, edges) {
+  const { mappedEdges, DOTToIdMap, idToDOTMap } = createMappings(nodes, edges);
```
[the remark this answers is not in the document]
Comment (aocenas) [reply public/app/plugins/panel/nodeGraph/layoutMsagl.worker.js:24]: I added a bit more explanation comment in the `createMappings` function

File: public/app/plugins/panel/nodeGraph/layoutMsagl.worker.js [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+
+  const dot = edgesToDOT(mappedEdges);
+  const graph = parseDot(dot);
+  const geomGraph = new GeomGraph(graph);
+  for (const e of graph.deepEdges) {
+    const gbc = new GeomEdge(e);
```
[the remark this answers is not in the document]
Comment (aocenas) [reply public/app/plugins/panel/nodeGraph/layoutMsagl.worker.js:30]: Removed

File: public/app/plugins/panel/nodeGraph/layoutMsagl.worker.js [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+}
+
+function edgesToDOT(edges) {
+  return toDOT(edges, 'rankdir="LR"; TBbalance="min"', '[ minlen=3 ]');
+}
+
+function nodesDOT(edges) {
+  let dot = '';
+  const visitedNodes = new Set();
+  // TODO: height/width for default sizing but nodes can have variable size now
```
[the remark this answers is not in the document]
Comment (aocenas) [reply public/app/plugins/panel/nodeGraph/layoutMsagl.worker.js:186]: Yeah this is needed also for the node overlap that you mentioned. But it's a bit older issue and not tied to this change so would try to fix it separately.

File: scripts/webpack/webpack.common.js [hunk 4]
```diff
@@ -119,6 +119,12 @@ module.exports = {
         test: /(unicons|mono|custom)[\\/].*\.svg$/,
         type: 'asset/source',
       },
+      {
+        test: /\.m?js$/,
+        resolve: {
+          fullySpecified: false,
+        },
+      },
```
[the remark this answers is not in the document]
Comment (aocenas) [reply scripts/webpack/webpack.common.js:127]: Added comment but without it the msagl library cannot be used. Without this webpack requires for module style js files to import with fully qualified names (including the extension) which the msagl library does not do.

**Review by joey-grafana — APPROVED** [review]
LGTM.. nice work! 👍 

Noticed one thing which is not a blocker. Errors in console caused by selecting `response_medium`. In the screenshot you can see a floating arrow (not connected to any node) around the center of the screen which is prob a clue!

Screenshot 2024-01-02 at 18 18 04

**Review by fabrizio-grafana — APPROVED** [review]

**Comment by drewcorlin1** [comment]
Is it possible to make this layout configurable? The "layered" layout is pretty much useless in my system where almost all services are called from a single ingress gateway. Same comment as https://github.com/grafana/grafana/issues/49988#issuecomment-2299364808

**Comment by drewcorlin1** [comment]
I poked around the code a bit and realized I can override the setting with `localStorage.setItem('grafana.featureToggles', 'nodeGraphDotLayout=true');` and that graph looks much better in my case. Is it possible to save this setting in my dashboard?

**Comment by joey-grafana** [comment]
Yes as you have noticed you can enable the `nodeGraphDotLayout` feature toggle but this is only enabled up to 500 nodes for performance reasons. Feature toggles are not permanent and in the future we will remove this so it will be enabled by default. 

> Is it possible to make this layout configurable? The "layered" layout is pretty much useless 

Here you say the layered layout is pretty much useless, yet that is what you have enabled by setting `nodeGraphDotLayout` to `true` - if you could clarify this that would be helpful.

**Comment by drewcorlin1** [comment]
Ah sorry I had the layout types backwards. The default layout was stacking almost every service in a vertical line but the layered was much more useful.

Here's a screenshot of what I was seeing before and after making the change (with service names redacted)

redacted-vertical

redacted-layered

**Comment by marwin1991** [comment]
@drewcorlin1 what have you changed to fix it?

**Comment by drewcorlin1** [comment]
I'm installing grafana via the kube-prometheus-stack helm chart, and I set `grafana."grafana.ini".feature_toggles.nodeGraphDotLayout` to true in my helm values

**Pull request merged**